cerealize/cerealize.h

```diff
@@ -225,37 +265,112 @@ class GenericCerealFileParser {
+  CerealBinaryFileParser(CerealBinaryFileParser&&) = delete;
+  void operator=(CerealBinaryFileParser&&) = delete;
 
   std::ifstream fi_;
-  typename CerealStreamType<T_CEREAL_FORMAT>::Input si_;
+  cereal::BinaryInputArchive si_;
 };
+
+
+// `CerealJSONFileParser` de-cereal-izes records from JSON file given their type 
+// and passes them over to `T_PROCESSOR`. Each line in the file expected to be a
+// full JSON record for one entry.
 template <typename T_ENTRY>
-using CerealFileParser = GenericCerealFileParser<T_ENTRY, CerealFormat::Default>;
+class CerealJSONFileParser {
+ public:
+  explicit CerealJSONFileParser(const std::string& filename) : fi_(filename) {}
+
+  // `Next` calls `T_PROCESSOR::operator()(const T_ENTRY&)` for the next entry, or returns false.
+  template <typename T_PROCESSOR>
+  bool Next(T_PROCESSOR&& processor) {
+    std::unique_ptr<T_ENTRY> entry = GetNextEntry();
+    if (entry) {
+      processor(*entry.get());
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // `NextWithDispatching` calls `T_PROCESSOR::operator()(const T_ACTUAL_ENTRY_TYPE&)`
+  // for the next entry, or returns false.
+  // Note that the actual entry type is being used in the calling method signature.
+  // This also makes it implausible to pass in a lambda here, thus the parameter is only passed by reference.
+  // In order for RTTI dispatching to work, class T_PROCESSOR should define two type:
+  // 1) BASE_TYPE: The type of the base entry to de-serialize from the stream, and
+  // 2) DERIVED_TYPE_LIST: An std::tuple<TYPE1, TYPE2, TYPE3, ...> of all the types that have to be matched.
+  template <typename T_PROCESSOR>
+  bool NextWithDispatching(T_PROCESSOR& processor) {
+    std::unique_ptr<T_ENTRY> entry = GetNextEntry();
+    if (entry) {
+      bricks::rtti::RuntimeTupleDispatcher<typename T_PROCESSOR::BASE_TYPE,
+                                           typename T_PROCESSOR::DERIVED_TYPE_LIST>::DispatchCall(*entry.get(),
+                                                                                                  processor);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+ private:
+  CerealJSONFileParser() = delete;
+  CerealJSONFileParser(const CerealJSONFileParser&) = delete;
+  void operator=(const CerealJSONFileParser&) = delete;
+  CerealJSONFileParser(CerealJSONFileParser&&) = delete;
+  void operator=(CerealJSONFileParser&&) = delete;
+
+  std::unique_ptr<T_ENTRY> GetNextEntry() {
+    std::string line;
+    if (std::getline(fi_, line)) {
+      std::istringstream iss(line);
+      std::unique_ptr<T_ENTRY> entry;
+      cereal::JSONInputArchive si_(iss);
+      si_(entry);
+      return std::move(entry);
+    } else {
+      return nullptr;
+    }
+  }
+
+  std::ifstream fi_;
+};
+
+template <typename T_ENTRY, CerealFormat>
+struct CerealFileParser {};
+
+template <typename T_ENTRY>
+struct CerealFileParser<T_ENTRY, CerealFormat::Binary> : CerealBinaryFileParser<T_ENTRY> {
```

Comment:
I'd prefer a `using`-based solution: https://gist.github.com/dkorolev/ce687e33b642e2de0ce3

As a small benefit, other constructors will get supported automatically upon being added to implementation/base class(es).